### PR TITLE
feat: support subindex on ExactNNSearch

### PR DESCRIPTION
**Pull request by maxwelljin**
Currently, the subindex feature is not available on exact in-memory NNIndexer. This PR aims to support nested documents for ExactNNIndexer

### CODE REVIEW

**Comment by maxwelljin** [comment]
This PR introduces support for sub-indexing in the exactNNSearch by enabling nested documents. Here's a summary of the changes:
1. Abstract class methods are utilized to create a parent_id / initialization for each nested document. 
2. The backend stores everything in the new_schema type, including the parent_id, but user need raw type. To handle this, the PR includes helper methods to convert the raw type back into the original type that the user inputted to the Indexer.
3. To ensure that the data type internally stored in the backend is not changed during casting, a copy method is included. The built-in document copy method does not support PyTorch tensors, so a new shallow copy method was created and tested.
4. The PR also includes test code, which is similar to the tests for other backends.
These changes aim to enhance the functionality of the exactNNSearch by supporting nested documents.

**Review by JoanFM — CHANGES_REQUESTED** [review]
what happens when I call persist, does it persist and reload all the subindices?

**Comment by JoanFM** [comment]
> > 
> 
> I'll take a look on that. All doc contents should be persist. If the load_binary function call the initialization, it should reconstruct subindices. I'll try some examples

And test it

**Comment by maxwelljin** [comment]
This new commit supports the persist storage on subindex document and includes the relevant test.

File: tests/index/in_memory/test_persist_data.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    list_tens: NdArray[20]
+
+
+class MyDoc(BaseDoc):
+    docs: DocList[SimpleDoc]
+    list_docs: DocList[ListDoc]
+    my_tens: NdArray[30]
+
+
+def test_subindex_index(tmp_path):
```
Comment (JoanFM) [inline tests/index/in_memory/test_persist_data.py:28]: I think we should also check that find and find_subindex work

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/index/in_memory/test_persist_data.py (hunk 1 above)
Comment (maxwelljin) [reply tests/index/in_memory/test_persist_data.py:28]: Sure, I'll add that

**Review by JoanFM — APPROVED** [review]

**Pull request merged**